### PR TITLE
feat(e2e): label and annotation

### PR DESCRIPTION
**Pull request by universal-itengineer**
## Description

Add e2e tests to check label and annotation on resources (pod, vm, vmi, etc.)
## Why do we need it, and what problem does it solve?

It is a part of task for automate e2e tests
## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: tests/e2e/kubectl/resource.go [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+	ResourcePod             Resource = "pod"
+	ResourceKubevirtVM      Resource = "virtualmachines.x.virtualization.deckhouse.io"
+	ResourceKubevirtVMI     Resource = "virtualmachineinstances.x.virtualization.deckhouse.io"
+	ResourceVM              Resource = "virtualmachine.virtualization.deckhouse.io"
+	ResourceVMIPClaim       Resource = "virtualmachineipaddressclaims.virtualization.deckhouse.io"
+	ResourceVMIPLeas        Resource = "virtualmachineipaddressleases.virtualization.deckhouse.io"
+	ResourceLabelKey        Resource = "os"
+	ResourceLabelValue      Resource = "ubuntu"
+	ResourceAnnotationKey   Resource = "test-annotation"
+	ResourceAnnotationValue Resource = "true"
```
Comment (yaroslavborbat) [inline tests/e2e/kubectl/resource.go:15]: These constants are not relevant to this package. And it's not a Resource. Move it to the test package

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: tests/e2e/vm_label_annotation_test.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+		getPodCMD := "get pod --no-headers -o custom-columns=':metadata.name'"
+		subCMD := fmt.Sprintf("-n %s %s | grep %s", namespace, getPodCMD, resourceName)
+		podCMD := kubectl.RawCommand(subCMD, ShortWaitDuration)
+		podName := strings.TrimSuffix(podCMD.StdOut(), "\n")
+		return podName
+	}
+
+	getRecourseLabel := func(resourceType kc.Resource, resourceName string) *executor.CMDResult {
+		label := fmt.Sprintf("jsonpath='{.metadata.labels.%s}'", labelName)
+		stdout := kubectl.GetResource(resourceType, resourceName, kc.GetOptions{
```
Comment (yaroslavborbat) [inline tests/e2e/vm_label_annotation_test.go:35]: name "stdout" is not correct.  Maybe res, result, cmdResult?

File: tests/e2e/vm_label_annotation_test.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+				subCMD := fmt.Sprintf("-n %s label vm %s %s=%s", conf.Namespace, vm.Name, labelName, labelValue)
+
+				res := kubectl.RawCommand(subCMD, ShortWaitDuration)
+				Expect(res.Error()).NotTo(HaveOccurred())
+			})
+		})
+		Describe("Check label on resource", func() {
+			It("VM", func() {
+				res := getRecourseLabel(kc.ResourceVM, vm.Name)
+				Expect(res.StdOut()).To(Equal(labelValue))
```
Comment (yaroslavborbat) [inline tests/e2e/vm_label_annotation_test.go:112]: you are not handling errors.
you can do it like this
`Expect(res.Error()).NotTo(HaveOccurred(), "failed to get vm %s.\n%s", vm.Name, res.StdErr())`

File: tests/e2e/vm_label_annotation_test.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+			})
+			It("KVVM", func() {
+				res := getRecourseLabel(kc.ResourceKubevirtVM, vm.Name)
+				Expect(res.StdOut()).To(Equal(labelValue))
```
Comment (yaroslavborbat) [inline tests/e2e/vm_label_annotation_test.go:116]: you are not handling errors

File: tests/e2e/vm_label_annotation_test.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+			})
+			It("KVVMI", func() {
+				res := getRecourseLabel(kc.ResourceKubevirtVMI, vm.Name)
+				Expect(res.StdOut()).To(Equal(labelValue))
```
Comment (yaroslavborbat) [inline tests/e2e/vm_label_annotation_test.go:120]: you are not handling errors.

File: tests/e2e/vm_label_annotation_test.go (hunk 3 above)
Comment (yaroslavborbat) [reply tests/e2e/vm_label_annotation_test.go:112]: you can add error handling to your functions getResourceAnnotation, getRecourseLabel, etc

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: tests/e2e/vm_label_annotation_test.go (hunk 4 above)
Comment (universal-itengineer) [reply tests/e2e/vm_label_annotation_test.go]: Added Expect(res.Error()).NotTo(HaveOccurred())

**Review by yaroslavborbat — APPROVED** [review]

File: tests/e2e/Taskfile.yaml [hunk 6]
```diff
@@ -28,3 +33,35 @@ tasks:
...
+      - ginkgo
+    cmds:
+      - |
+        {{if .TEST }}
+          export KUBECONFIG={{.KUBECONFIG}}
+          export E2E_CLUSTERTRANSPORT_KUBECONFIG={{.E2E_CLUSTERTRANSPORT_KUBECONFIG}}
+          ginkgo --focus "{{ .TEST }}" -v
+        {{else}}
+          echo "Test not passed"
+          echo 'Example: TEST="Label and Annotation" task run_one'
```
Comment (diafour) [inline tests/e2e/Taskfile.yaml:63]: Shouldn't it fail if "not passed"?
```suggestion
          echo 'Example: TEST="Label and Annotation" task run_one'
          exit 1
```

File: tests/e2e/vm_label_annotation_test.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+import (
+	"fmt"
+	"github.com/deckhouse/virtualization/tests/e2e/executor"
+	kc "github.com/deckhouse/virtualization/tests/e2e/kubectl"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"io/fs"
+	"path/filepath"
+	"strings"
```
Comment (diafour) [inline tests/e2e/vm_label_annotation_test.go:11]:
```suggestion
	"fmt"
	"io/fs"
	"path/filepath"
	"strings"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	"github.com/deckhouse/virtualization/tests/e2e/executor"
	kc "github.com/deckhouse/virtualization/tests/e2e/kubectl"
```

File: tests/e2e/vm_label_annotation_test.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+				res := getRecourseLabel(kc.ResourcePod, pod)
+				Expect(res.Error()).NotTo(HaveOccurred(), "failed to get pod %s.\n%s", vm.Name, res.StdErr())
+				Expect(res.StdOut()).To(Equal(labelValue))
+			})
+		})
+
+		Describe(fmt.Sprintf("Remove label %s=%s", labelName, labelValue), func() {
+
+			It("Label was removed", func() {
+				Expect(err).To(BeNil())
```
Comment (diafour) [inline tests/e2e/vm_label_annotation_test.go:136]: Is it needed?

File: tests/e2e/Taskfile.yaml (hunk 6 above)
Comment (universal-itengineer) [reply tests/e2e/Taskfile.yaml:63]: exit 1 - not necessary 
`echo "Test not passed"` changed to `echo "Specify test to run"`

File: tests/e2e/vm_label_annotation_test.go [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+			break
+		}
+	}
+	var vm virtualMachine
+
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstruct.Object, &vm); err != nil {
+		return nil, err
+	}
+	return &vm, nil
+}
```
Comment (yaroslavborbat) [inline tests/e2e/vm_label_annotation_test.go:44]: This function has been copied. You can use GetVMFromManifest

File: tests/e2e/vm_label_annotation_test.go [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+	"github.com/deckhouse/virtualization/tests/e2e/executor"
+	"github.com/deckhouse/virtualization/tests/e2e/helper"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	kc "github.com/deckhouse/virtualization/tests/e2e/kubectl"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type virtualMachine struct {
```
Comment (yaroslavborbat) [inline tests/e2e/vm_label_annotation_test.go:21]: This structure is not needed

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

**Review by yaroslavborbat — APPROVED** [review]

File: tests/e2e/vm_label_annotation_test.go [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+			Output:    annotation,
+			Namespace: conf.Namespace,
+		})
+		return cmdResult
+	}
+
+	WaitVmStatus := func(name, phase string) {
+		GinkgoHelper()
+		WaitResource(kc.ResourceVM, name, "jsonpath={.status.phase}="+phase, LongWaitDuration)
+	}
```
Comment (diafour) [inline tests/e2e/vm_label_annotation_test.go:56]: Why not create an additional helper WaitResourceStatusPhase?

File: tests/e2e/util_test.go [hunk 12]
```diff
@@ -7,7 +7,12 @@ import (
 	kc "github.com/deckhouse/virtualization/tests/e2e/kubectl"
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
```
Comment (diafour) [inline tests/e2e/util_test.go:15]: Unsorted imports. Hm.. we need linter for tests/e2e.

File: tests/e2e/vm_label_annotation_test.go (hunk 11 above)
Comment (universal-itengineer) [reply tests/e2e/vm_label_annotation_test.go:56]: Not in this PR

File: tests/e2e/util_test.go (hunk 12 above)
Comment (universal-itengineer) [reply tests/e2e/util_test.go:15]: Not in this PR

**Pull request merged**